src/Type/Doctrine/Query/QueryResultDynamicReturnTypeExtension.php

```diff
@@ -126,10 +142,17 @@ private function getMethodReturnTypeForHydrationMode(
 					$queryResultType
 				);
 			default:
-				return new ArrayType(
-					new MixedType(),
-					$queryResultType
-				);
+				return $queryIndexType instanceof MixedType
+					? AccessoryArrayListType::intersectWith(
+						TypeCombinator::intersect(
+							new ArrayType(new IntegerType(), $queryResultType),
+							...TypeUtils::getAccessoryTypes($queryResultType)
+						)
+					)
+					: new ArrayType(
+						$queryIndexType,
```

Comment:
This can really be only string or int. We should handle nullable values, objects, booleans etc. 

**Or** we could simplify this MR not to try handle those cases and just distinguish list / array whether indexBy was used or not.

Reply:
I'd start with just distinguishing list / array while having the new generic type of Query as seen in this MR to be able to improve it later.